provider/openstack/firewaller.go

```diff
@@ -718,18 +719,32 @@ func (c *neutronFirewaller) openPortsInGroup(nameRegExp string, portRanges []net
 	return nil
 }
 
-// ruleMatchesPortRange checks if supplied nova security group rule matches the port range
-func ruleMatchesPortRange(rule neutron.SecurityGroupRuleV2, portRange network.PortRange) bool {
-	if rule.IPProtocol == nil || *rule.PortRangeMax == 0 || *rule.PortRangeMin == 0 {
+// secGroupMatchesIngressRule checks if supplied nova security group rule matches the ingress rule
+func secGroupMatchesIngressRule(secGroupRule neutron.SecurityGroupRuleV2, rule network.IngressRule) bool {
+	if secGroupRule.IPProtocol == nil || *secGroupRule.PortRangeMax == 0 || *secGroupRule.PortRangeMin == 0 {
 		return false
 	}
-	return *rule.IPProtocol == portRange.Protocol &&
-		*rule.PortRangeMin == portRange.FromPort &&
-		*rule.PortRangeMax == portRange.ToPort
+	portsMatch := *secGroupRule.IPProtocol == rule.Protocol &&
+		*secGroupRule.PortRangeMin == rule.FromPort &&
+		*secGroupRule.PortRangeMax == rule.ToPort
+	if !portsMatch {
+		return false
+	}
+	// The ports match, so if the security group RemoteIPPrefix matches *any* of the
+	// rule's source ranges, then that's a match.
+	if len(rule.SourceRanges) == 0 {
+		return secGroupRule.RemoteIPPrefix == "" || secGroupRule.RemoteIPPrefix == "0.0.0.0/0"
+	}
+	for _, r := range rule.SourceRanges {
```

Comment:
The openstack security group only has a single rule (RemoteIPPrefix). So all we need to do is see if the RemoteIPPrefix matches any of the source CIDRs in the ingress rule.